vfio/src/vfio.rs

```diff
@@ -0,0 +1,596 @@
+    GroupGetDeviceFD,
+    CreateVfioKvmDevice(io::Error),
+    KvmSetDeviceAttr(io::Error),
+    VfioDeviceGetInfo,
+    VfioDeviceGetRegionInfo,
+    InvalidPath,
+    IommuDmaMap,
+    IommuDmaUnmap,
+    VfioDeviceGetIrqInfo,
+    VfioDeviceSetIrq,
+}
+pub type Result<T> = std::result::Result<T, VfioError>;
+
+impl fmt::Display for VfioError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            VfioError::OpenContainer(e) => write!(f, "failed to open /dev/vfio/vfio container: {}", e),
+            VfioError::OpenGroup(e) => write!(f, "failed to open /dev/vfio/$group_num group: {}", e),
+            VfioError::GetGroupStatus => write!(f, "failed to get Group Status"),
+            VfioError::GroupViable => write!(f, "group is inviable"),
+            VfioError::VfioApiVersion => write!(f, "vfio API version doesn't match with VFIO_API_VERSION defined in vfio_sys/srv/vfio.rs"),
+            VfioError::VfioType1V2 => write!(f, "container dones't support VfioType1V2 IOMMU driver type"),
+            VfioError::GroupSetContainer => write!(f, "failed to add vfio group into vfio container"),
+            VfioError::ContainerSetIOMMU => write!(f, "failed to set container's IOMMU driver type as VfioType1V2"),
+            VfioError::GroupGetDeviceFD => write!(f, "failed to get vfio device fd"),
+            VfioError::CreateVfioKvmDevice(e) => write!(f, "failed to create KVM vfio device: {}", e),
+            VfioError::KvmSetDeviceAttr(e) => write!(f, "failed to set KVM vfio device's attribute: {}", e),
+            VfioError::VfioDeviceGetInfo => write!(f, "failed to get vfio device's info or info doesn't match"),
+            VfioError::VfioDeviceGetRegionInfo => write!(f, "failed to get vfio device's region info"),
+            VfioError::InvalidPath => write!(f,"invalid file path"),
+            VfioError::IommuDmaMap => write!(f, "failed to add guest memory map into iommu table"),
+            VfioError::IommuDmaUnmap => write!(f, "failed to remove guest memory map from iommu table"),
+            VfioError::VfioDeviceGetIrqInfo => write!(f, "failed to get vfio device irq info"),
+            VfioError::VfioDeviceSetIrq => write!(f, "failed to set vfio deviece irq"),
+        }
+    }
+}
+
+struct VfioContainer {
+    container: File,
+}
+
+const VFIO_API_VERSION: u8 = 0;
+impl VfioContainer {
+    fn new() -> Result<Self> {
+        let container = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open("/dev/vfio/vfio")
+            .map_err(|e| VfioError::OpenContainer(e))?;
+
+        Ok(VfioContainer { container })
+    }
+
+    fn get_api_version(&self) -> i32 {
+        // Safe as file is vfio container fd and ioctl is defined by kernel.
+        unsafe { ioctl(self, VFIO_GET_API_VERSION()) }
+    }
+
+    fn check_extension(&self, val: u32) -> bool {
+        if val != VFIO_TYPE1_IOMMU && val != VFIO_TYPE1v2_IOMMU {
+            panic!("IOMMU type error");
```

Comment:
Do we really want to panic here? A nice Error return would be preferable IMO.

Reply:
Done.